drv/i2c-devices/src/sbrmi.rs

```diff
@@ -0,0 +1,339 @@
+    }
+
+    fn read_reg(&self, reg: Register) -> Result<u8, Error> {
+        match self.device.read_reg::<u8, u8>(reg.into()) {
+            Ok(buf) => Ok(buf),
+            Err(code) => Err(Error::BadRegisterRead { reg, code }),
+        }
+    }
+
+    pub fn nthreads(&self) -> Result<u8, Error> {
+        self.read_reg(Register::ThreadNumber)
+    }
+
+    pub fn enabled(&self) -> Result<[u8; 16], Error> {
+        //
+        // The enabled bits are found in three different banks across the
+        // register space.  The banks are asymmetric (which is slightly
+        // annoying), but -- unlike the alert bits -- at least the enabled
+        // bits are stored contiguously by thread...
+        //
+        let mut rval = [0u8; 16];
+        let mut roffs = 0;
+
+        for (base, banksize) in &[(0x4, 2), (0x8, 6), (0x43, 8)] {
+            for offs in 0..*banksize {
+                rval[roffs + offs] = self.read_reg(Register::Enabled {
+                    base: *base,
+                    offset: offs as u8,
+                })?;
+            }
+
+            roffs += banksize;
+        }
+
+        Ok(rval)
+    }
+
+    pub fn alert(&self) -> Result<[u8; 16], Error> {
+        //
+        // Have you ever played 52-bit Pickup?  Unlike the enabled bits, the
+        // alert bits are smeared and interleaved across the register space.
+        // Each byte has 4 bits, with each bit representing the alert status
+        // of a thread; here is the mapping of registers to threads:
+        //
+        //   REGISTER  DESCRIPTION
+        //   0x10      MceStat[3:0] = Threads[48,32,16,0]
+        //   0x11      MceStat[3:0] = Threads[49,33,17,1]
+        //   0x12      MceStat[3:0] = Threads[50,34,18,2]
+        //   0x13      MceStat[3:0] = Threads[51,35,19,3]
+        //   0x14      MceStat[3:0] = Threads[52,36,20,4]
+        //   0x15      MceStat[3:0] = Threads[53,37,21,5]
+        //   0x16      MceStat[3:0] = Threads[54,38,22,6]
+        //   0x17      MceStat[3:0] = Threads[55,39,23,7]
+        //   0x18      MceStat[3:0] = Threads[56,40,24,8]
+        //   0x19      MceStat[3:0] = Threads[57,41,25,9]
+        //   0x1A      MceStat[3:0] = Threads[58,42,26,10]
+        //   0x1B      MceStat[3:0] = Threads[59,43,27,11]
+        //   0x1C      MceStat[3:0] = Threads[60,44,28,12]
+        //   0x1D      MceStat[3:0] = Threads[61,45,29,13]
+        //   0x1E      MceStat[3:0] = Threads[62,46,30,14]
+        //   0x1F      MceStat[3:0] = Threads[63,47,31,15]
+        //   0x50      MceStat[3:0] = Threads[112,96,80,64]
+        //   0x51      MceStat[3:0] = Threads[113,97,81,65]
+        //   0x52      MceStat[3:0] = Threads[114,98,82,66]
+        //   0x53      MceStat[3:0] = Threads[115,99,83,67]
+        //   0x54      MceStat[3:0] = Threads[116,100,84,68]
+        //
+        let mut rval = [0u8; 16];
+        let banksize = 16usize;
+
+        for (bank, base) in [0x10, 0x50].iter().enumerate() {
```

Comment:
Is it harmless to read extra registers that aren't in the table above (`0x55-0x5F`)?

If not, we could do something like
```suggestion
        for (bank, (base, banksize0)) in
            [(0x10, 16), (0x50, 5)].iter().enumerate()
        {
```

Reply:
Ha -- the comment was incomplete!  I had neglected to include the second page of the table; thanks for spotting that!